worlds/yugioh06/opponents.py

```diff
@@ -0,0 +1,257 @@
+        OpponentData(19, "Elemental Hero Erikshieler", ["Quick-Finish"], 4, 5, 6639, 8019,
+                     "deck/LV4_Ehero.ydc\x00\x00\x00\x00", 3),
+        # Tier 5
+        OpponentData(20, "Raviel, Lord of Phantasms", [], 5, 1, 6565, 8020, "deck/LV5_ravieru.ydc\x00\x00", 4),
+        OpponentData(21, "Horus the Black Flame Dragon LV8", [], 5, 2, 6100, 8021, "deck/LV5_horus.ydc\x00\x00\x00\x00",
+                     4),
+        OpponentData(22, "Stronghold", [], 5, 3, 6153, 8022, "deck/LV5_gadget.ydc\x00\x00\x00", 5),
+        OpponentData(23, "Sacred Phoenix of Nephthys", [], 5, 4, 6236, 8023, "deck/LV5_nephthys.ydc\x00", 6),
+        OpponentData(24, "Cyber End Dragon", ["Goal"], 5, 5, 6397, 8024, "deck/LV5_cyber.ydc\x00\x00\x00\x00", 7),
+    ]
+    world = multiworld.worlds[player]
+    if not randomize:
+        return opponents_table
+    opponents = opponents_table + challenge_opponents
+    start = world.random.choice([o for o in opponents if o.tier == 1 and len(o.additional_info) == 0])
+    opponents.remove(start)
+    goal = world.random.choice([o for o in opponents if "Goal" in o.campaign_info])
+    opponents.remove(goal)
+    world.random.shuffle(opponents)
+    chosen_ones = opponents[:23]
+    for item in (multiworld.precollected_items[player]):
+        if item.name in tier_1_opponents:
+            # convert item index to opponent index
+            chosen_ones.insert(item_to_index[item.name] - item_to_index["Campaign Tier 1 Column 1"], start)
+            continue
+    chosen_ones.append(goal)
+    tier = 1
+    column = 1
+    recreation = []
+    for opp in chosen_ones:
+        recreation.append(OpponentData(opp.id, opp.name, opp.campaign_info, tier, column, opp.card_id,
+                                       opp.deck_name_id, opp.deck_file, opp.difficulty))
+        column += 1
+        if column > 5:
+            column = 1
+            tier += 1
+
+    return recreation
+
+
+def get_opponent_locations(opponent: OpponentData) -> Dict[str, Optional[Union[str, int]]]:
+    location = {opponent.name + " Beaten": "Tier " + str(opponent.tier) + " Beaten"}
+    if opponent.tier > 4 and opponent.column != 5:
+        name = "Campaign Tier 5: Column " + str(opponent.column) + " Win"
+        # return a int instead so a item can be placed at this location later
+        location[name] = special[name]
+    for info in opponent.campaign_info:
+        location[opponent.name + "-> " + info] = info
+    return location
+
+
+def get_opponent_condition(opponent: OpponentData, unlock_item: str, unlock_amount: int, player: int,
+                           is_challenge: bool):
+    return lambda state: (
+            ((is_challenge and state.has(unlock_item, player, unlock_amount))
+             or (not is_challenge and state.has_group(unlock_item, player, unlock_amount)))
+            and yugioh06_difficulty(state, player, opponent.difficulty)
+            and state.has_all(opponent.additional_info, player)
+    )
```

Comment:
This would actually be slightly faster (and imo slightly more readable too? Even if it's longer) in practice if you did it like this:

```
if is_challenge:
    return lambda state: (
        state.has(unlock_item, player, unlock_amount)
        and yugioh06_difficulty(state, player, opponent.difficulty)
        and state.has_all(opponent.additional_info, player)
    )
else:
    return lambda state: (
        state.has_group(unlock_item, player, unlock_amount)
        and yugioh06_difficulty(state, player, opponent.difficulty)
        and state.has_all(opponent.additional_info, player)
    )
```

*Low priority but I think it's worth it*